azurerm/internal/services/kusto/resource_arm_kusto_eventgrid_data_connection.go

```diff
@@ -0,0 +1,284 @@
+package kusto
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/services/kusto/mgmt/2019-05-15/kusto"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/features"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmKustoEventGridDataConnection() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmKustoEventGridDataConnectionCreateUpdate,
+		Read:   resourceArmKustoEventGridDataConnectionRead,
+		Update: resourceArmKustoEventGridDataConnectionCreateUpdate,
+		Delete: resourceArmKustoEventGridDataConnectionDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(60 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Update: schema.DefaultTimeout(60 * time.Minute),
+			Delete: schema.DefaultTimeout(60 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateAzureRMKustoDataConnectionName,
```

Comment:
IMHO the validation should be done against ARM calling dataConnectionValidation and not locally. Validation can be changed in the server-side and it doesn't make sense to duplicate it in client code and to have to change it. This command is valid to all other validate calls in this file

Reply:
Doing a check here is actually quite helpful as it prevents Terraform from stopping mid execution because a validation check had failed. This check prevents anything from being created until the validation succeeds. Also not all APIs gracefully handle validation or incorrect data so we err on the side of caution. The APIs might not offer any information about why a value is wrong so this check ensures there is an error a user can take action on. Lastly, this check happens before the api is every hit which ensures these issues are discovered faster.

Reply:
Yes, I understand the trade-offs. The pro is that the validation is done client-side and it fails fast in case the input is not valid. It also reduces the number of calls against the API. The con is that (some of) the validation logic duplicated and the same code exists in the server-side - if it changes there, someone needs to remember to update the client-side validation. In case the API handles the validation gracefully and offers information about why the value is wrong, do you find it reasonable to validate by calling the validate API? or is it preferable to deal with the changes when they occur?

Reply:
@ilayrn, at the moment we do all validation locally and do not reach out to the API.